x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/scheduling/TransformScheduler.java

```diff
@@ -0,0 +1,236 @@
+
+/**
+ * {@link TransformScheduler} class is responsible for scheduling transform tasks according to their configured frequency as well as
+ * retrying policy.
+ */
+public final class TransformScheduler {
+
+    /**
+     * {@link Event} record encapsulates data that might be useful to the listener when the listener is being triggered.
+     */
+    public record Event(String transformId, long scheduledTime, long triggeredTime) {
+        @Override
+        public String toString() {
+            return new StringBuilder("Event[").append("transformId=")
+                .append(transformId)
+                .append(",scheduledTime=")
+                .append(Instant.ofEpochMilli(scheduledTime))
+                .append(",triggeredTime=")
+                .append(Instant.ofEpochMilli(triggeredTime))
+                .append("]")
+                .toString();
+        }
+    }
+
+    /**
+     * {@link Listener} interface allows receiving events about the scheduled task being triggered.
+     */
+    public interface Listener {
+        /**
+         * Called whenever scheduled task is being triggered.
+         *
+         * @param event event structure containing data that might be useful to the listener
+         */
+        void triggered(Event event);
+    }
+
+    private static final Logger logger = LogManager.getLogger(TransformScheduler.class);
+
+    private final Clock clock;
+    private final ThreadPool threadPool;
+    private final TimeValue schedulerFrequency;
+    private final TransformScheduledTaskQueue scheduledTasks;
+    /**
+     * Prevents two concurrent invocations of the "processScheduledTasks" method.
+     *
+     * Set to {@code true} before processing
+     * Set to {@code false} after processing (doesn't matter whether successful or not).
+     */
+    private final AtomicBoolean isProcessingActive;
+
+    public TransformScheduler(Clock clock, ThreadPool threadPool, Settings settings) {
+        this.clock = Objects.requireNonNull(clock);
+        this.threadPool = Objects.requireNonNull(threadPool);
+        this.schedulerFrequency = Transform.SCHEDULER_FREQUENCY.get(settings);
+        this.scheduledTasks = new TransformScheduledTaskQueue();
+        this.isProcessingActive = new AtomicBoolean();
+    }
+
+    /**
+     * Starts the processing.
+     * Method {@code processScheduledTasks} is invoked according to its configured {@code schedulerFrequency}.
+     */
+    public void start() {
+        threadPool.scheduleWithFixedDelay(this::processScheduledTasks, schedulerFrequency, ThreadPool.Names.GENERIC);
+    }
+
+    // Visible for testing
+    void processScheduledTasks() {
+        // Prevent two concurrent invocations of the "processScheduledTasks" method
+        if (isProcessingActive.compareAndSet(false, true) == false) {
+            return;
+        }
+        logger.trace("Processing scheduled tasks started");
+        Instant processingStarted = clock.instant();
```

Comment:
If `processingStarted` is only required for trace logging perhaps only create it if trace is enabled;

```
Instant processingStarted = null;
if (logger.isTraceEnabled()) {
    processingStarted = clock.instant();
}
```

I honestly don't know if this makes a difference as the compiler might optimise it away

Reply:
Done.